velox/functions/prestosql/FromUtf8.cpp

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/StringWriter.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/string/StringImpl.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+class FromUtf8Function : public exec::VectorFunction {
+ public:
+  void apply(
```

Comment:
why can't this be a simple function?

Reply:
In most cases, we have all-ASCII input or all-valid UTF-8 input. In these cases we can avoid copying input and return it as is. But if there is an invalid sequence of bytes, then we need to copy data. Simple function doesn't allow for implementing these optimizations.